PyInstaller/depend/analysis.py

```diff
@@ -133,6 +133,68 @@ def _reset(self, user_hook_dirs):
         self._hooks_pre_safe_import_module = self._cache_hooks('pre_safe_import_module')
         self._hooks_pre_find_module_path = self._cache_hooks('pre_find_module_path')
 
+        # Search for run-time hooks in all hook directories.
+        self._available_rthooks = defaultdict(list)
+        for uhd in self._user_hook_dirs:
+            uhd_path = os.path.join(uhd, 'rthooks.dat')
+            try:
+                with compat.open_file(uhd_path, compat.text_read_mode,
+                                      encoding='utf-8') as f:
+                    available_rthooks = ast.literal_eval(f.read())
+            except FileNotFoundError:
+                # Ignore if this hook path doesn't have run-time hooks.
+                continue
+            except Exception as e:
+                logger.error('Unable to read run-time hooks from %s: %s' %
+                             (uhd_path, str(e)))
+                continue
+
+            # The expected data structure for a run-time hook file is::
+            #
+            #   {
+            #       str module_name1: [
+            #           str python_file_name1,
+            #           ...
+            #       ],
+            #       ...
+            #   }
+            #
+            # Check then merge this data structure, updating the file names to be absolute.
```

Comment:
Hmmm. This is an interesting point. If we issue a warning, that implies (at least to me) that the current data structure in `rthooks.dat` should be `Dict[str, str]`, not `Dict[str, List[str]]`, since there shouldn't be two hooks for a given module. Put another way, why does the current data structure provide a list of files names associated with a given module? Did we at some point expect to allow multiple rthooks for each module?

Reply:
One answer to my question above: the current `rthooks.dat` specifies multiple hooks for a given module (matplotlib). The new case is when multiple `rthooks.dat` files specify hooks for the same module.

- This would be OK if the hooks don't interfere. For example, one hook for a hypothetical satellite_tracking package would apply to the load_satellites module, including its geostationary satellite data, while another package would also include a hook for load_satellites to include low earth orbit satellites.
- This would fail if the hooks conflicted. If one satellite_tracking hook specified the value for g more precisely for earth, while another specified martian gravity, one would lose.

However, I don't see how PyInstaller could decide which case applies, It could potentially issue a warning if there are multiple run-time hooks added by differing `rthooks.dat` files, but this would take additional effort (recording the `rthooks.dat` file from which each hook was added, and changing the data structure to accommodate this additional information). Therefore, I suggest omitting this check.